src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -244,24 +244,35 @@ public boolean isNative() {
     @Override
     public final Optional<MemorySegment> asOverlappingSlice(MemorySegment other) {
         final AbstractMemorySegmentImpl that = (AbstractMemorySegmentImpl)Objects.requireNonNull(other);
-        if (overlaps(that)) {
-            final long offsetToThat = that.address() - this.address();
-            final long newOffset = offsetToThat >= 0 ? offsetToThat : 0;
-            return Optional.of(asSlice(newOffset, Math.min(this.byteSize() - newOffset, that.byteSize() + offsetToThat)));
+        if (overlaps(that) == 0) {
+            return Optional.empty();
         }
-        return Optional.empty();
+        final long offsetToThat = that.address() - this.address();
+        final long newOffset = offsetToThat >= 0 ? offsetToThat : 0;
+        return Optional.of(asSlice(newOffset, Math.min(this.byteSize() - newOffset, that.byteSize() + offsetToThat)));
     }
 
+    // Returns 1 if the regions overlap, otherwise 0.
     @ForceInline
-    boolean overlaps(AbstractMemorySegmentImpl that) {
+    int overlaps(AbstractMemorySegmentImpl that) {
         if (unsafeGetBase() == that.unsafeGetBase()) {  // both either native or the same heap segment
+
             final long thisStart = this.unsafeGetOffset();
-            final long thatStart = that.unsafeGetOffset();
             final long thisEnd = thisStart + this.byteSize();
+            final long thatStart = that.unsafeGetOffset();
             final long thatEnd = thatStart + that.byteSize();
-            return (thisStart < thatEnd && thisEnd > thatStart); //overlap occurs?
-        }
-        return false;
+
+            // The below computation is a branchless equivalent to
+            // `return (thisStart < thatEnd && thisEnd > thatStart)?1:0;`. Here is how:
```

Comment:
It would be pretty cool if C2 could do the optimization of branch-less computations like these.

Reply:
Yes. My thought as well. I will make some inquiries.